### ISSUE TITLE
Input focus issues

### ISSUE THREAD

**Issue by m-from-space** [renpy/renpy#5205]
Consider this:

```
default input_var1 = "hello"
default input_var2 = "world"

screen testscreen():
    vbox:
        input value VariableInputValue("input_var1")
        input value VariableInputValue("input_var2")
```

Not only is the first input field selected by default (which it shouldn't be), I also cannot get focus on the other input field. What is going on? I am so confused and couldn't find an explanation in the documentation.

Renpy Version: Latest Nightly Fix Python 3

**Comment by Gouvernathor** [comment]
As shown in [the documentation](https://www.renpy.org/doc/html/screen_actions.html#input-values), the `default=True` parameter is passed to the values you're creating. Since you can't focus several things, only the first one gets focused.
Then, to *change* the focusness of inputs, you need to trigger (somehow, presumably with a button and possibly by putting the input as the button's child so it's like a textbutton) the .Toggle (or .Enable or .Disable) action of the value.
How do you do that when the input value is generated directly in the input ? You don't. Put the values in default statements *in the screen* (or out if you prefer), and put the variable name in the input's value and the Toggle in the button's action.

Works as intended.

**Comment by m-from-space** [comment]
> Works as intended.
Oh dear. First of all thank you. But this was the most complex thing I had to figure out with Renpy over many years, lmao.

So you basically have to do this (which is ridiculous in my opinion):

```
default myvar = "123"

screen myscreen():
    default input_object = VariableInputValue("myvar", default=False)
    button:
        action input_object.Toggle()
        # you better give it a size if you want to be able to trigger it
        xysize (200, 30)
        input value input_object
```
I mean maybe it could be a nice addition if you just create an "input" that works as an input is to be expected: a textfield you can click into and type in stuff. o-o

On the other hand the documentation clearly needs an example here.

Thanks again though.

**Comment by Gouvernathor** [comment]
> On the other hand the documentation clearly needs an example here.

I found this trick in the file_slots screen of the default GUI. But maybe that recipe would gain more visibility in the docs, I don't know.

The trouble with an "all-in-one" input as you would have it, so an inputbutton really, is that it would be much harder to customize than what the current input allows. But we could share a recipe to make `inputbutton:` a thing using [renpy.register_sl_statement](https://nightly.renpy.org/doc/screen_python.html#renpy.register_sl_statement). I'm frosty about adding it to renpy itself though (it's not unseen of, since `label` and `textbutton` are little more than glorified wrappers over button o input, but I'm still not too keen on doing that).

Also, instead of giving it an xysize you can let it have a default text through the [`default` property of the input](https://nightly.renpy.org/doc/screens.html#screen-property-default) (not to be confused with the default parameter to the input*value*). But in that case you still have the trap of letting the player erase all characters in the input, which you have to countermand in some way.

**Comment by m-from-space** [comment]
> The trouble with an "all-in-one" input as you would have it, so an inputbutton really, is that it would be much harder to customize than what the current input allows.

I love it that Renpy allows you to do so many things by staying flexible like this. I'm just saying - like you - that there could be a default "inputfield" available for users as well as the current situation.

I have another problem with inputs and I don't want to open up a new issue so far, maybe you have an idea: Sometimes my inputs do not react to user keypress events. I use the "allowed" parameter defining allowed characters, for example "0123456789". Sometimes I click inside, but neither does it allow any characters, you also cannot delete ones. Moving to another input works, coming back sometimes works, sometimes not. There is something going on here. Is there a way to debug things?

**Comment by Gouvernathor** [comment]
You should pass default=False to all your input values, make sure your buttons display differently when enabled and disabled, and check the focusness of all your inputs. You can use the DisableAllInputValues action too.
Also, this kind of exploration is better done on the discord help channel than here on github.

**Comment by m-from-space** [comment]
> You should pass default=False to all your input values, make sure your buttons display differently when enabled and disabled, and check the focusness of all your inputs. You can use the DisableAllInputValues action too. Also, this kind of exploration is better done on the discord help channel than here on github.

I passed the `default=False` to all inputs, the input gets focussed and it seems also happening (quite often!) when not excluding any characters. Just now, it happened 5 times when I testet it. I click inside the input, the caret appears, but I cannot put anything inside, nor can I delete the input. I use DisableAllInputValues() as part of a dismiss and as part of a reset of the input fields, that works fine.

**Comment by Gouvernathor** [comment]
Hm, that sounds a bit more like a bug then, the caret should only appear when you can type.
Can you provide a short code reproducing the issue, and a consistent list of things to do to reproduce that state in the game ?

**Comment by m-from-space** [comment]
> Hm, that sounds a bit more like a bug then, the caret should only appear when you can type. Can you provide a short code reproducing the issue, and a consistent list of things to do (as a player) to reproduce that state in the game ?

Yeah, there's definitely something going on here. So here is the way my screen is setup (it's a configuration screen with 9 input fields for certain parameters).

Here is an example. I could reproduce the issue easily, by clicking inside an input, trying to put something inside or delete. If it works, switch to another input and try again. Only after 1 to 3 times, it doesn't work. Then it suddenly works again...

(By the way, it doesn't matter if the input objects are created using "default" or if they are screen variables. Same problems.)

```
default persistent.input_val_1 = "1024"
default persistent.input_val_2 = "5.5"
default persistent.input_val_3 = "-100"

define input_1 = VariableInputValue("persistent.input_val_1", default=False)
define input_2 = VariableInputValue("persistent.input_val_2", default=False)
define input_3 = VariableInputValue("persistent.input_val_3", default=False)

screen configuration():
    tag menu
    use game_menu(_("Configuration"), scroll="viewport"):
        grid 3 3:
            xspacing 40
            yspacing 10
            vbox:
                text "Input 1"
                button:
                    action input_1.Toggle()
                    xysize (200, 30)
                    input value input_1 length 6
            vbox:
                text "Input 2"
                button:
                    action input_2.Toggle()
                    xysize (200, 30)
                    input value input_2 length 6
            vbox:
                text "Input 3"
                button:
                    action input_3.Toggle()
                    xysize (200, 30)
                    input value input_3 length 6
```

**Comment by Gouvernathor** [comment]
The input values are not constant, since they enable and disable through time. They should not be created using `define`. That may be what's inducing errors for the screen language system. The reverse issue is expected if you use `$` inside the screen to create the input values, as they would reset each time. You should only use `default`, either outside the screen if you want the selectedness to persist when closing the menu and reopening it, or inside the screen if not.
Also, you should be using FieldInputValue instead of VariableInputValue.
I will do some tests myself later.

**Comment by m-from-space** [comment]
> The input values are not constant, since they enable and disable through time. They should not be created using `define`.

As I said, it doesn't matter. I also don't quite agree here, because it's not about the values, but the objects. You also know, that using "define" just tells Renpy it shouldn't care about saving it and that it will do stuff with it in init time. I actually think that "default" would be wrong here, since this is not something that Renpy needs to change during after_load or start. As I said, same issues with "default".

> The reverse issue is expected if you use `$` inside the screen to create the input values, as they would reset each time.

Yeah, I didn't do that. :)

> Also, you should be using FieldInputValue instead of VariableInputValue.

That's what I thought, but that doesn't work at all. Also in the documentation there is a specific example with "VariableInputValue()" used on persistent variables.

**Comment by Gouvernathor** [comment]
No, the define statement also marks the value as a constant, and the screens system saves some time by not recomputing values it considers constant. Creating the variable in an init python block would work better as far as the screen system is concerned.

**Comment by m-from-space** [comment]
Alright, so I did some more testing because of your suggestions.

* using FieldInputValues significantly reduces the occurance of the problem (but doesn't resolve it)
* using "define" works WAY better (meaning less often it's not editable)
* using "default" works worse (doesn't matter if global or screen variables)
* using an "init python" block works as bad as "default"

Thanks for keeping up with me, I hope you can reproduce the issue.

**Comment by HorBro38** [comment]
I'm not sure if this will help you @m-from-space , but this is how I allow for multiple inputs on one screen:

```py
frame:
        style "s_frame_stone"
        xysize (1310, 180)
        hbox:
          vbox:            
            if active_input == 1:
              input:
                length 25
                value FieldInputValue(currChar, "name")
            else:
              textbutton "[currChar.name]" action (SetVariable("active_input", 1), Function(text_saver))
            if active_input == 2:
                  input:
                    length 15
                    value FieldInputValue(currChar, "gender")
            else:
                  textbutton "[currChar.gender]" action (SetVariable("active_input", 2), Function(text_saver))
```
I basically track `active_input` to see which input should be valid at that time, and allow click on each to allow input.  It defaults to 0.  text_saver is a function that resets variables if the player enters nothing, simply so they will always have something they can click on:

```py
def text_saver():
    if currChar.name == "":
      currChar.name = "Character Name"
    if currChar.gender == "":
      currChar.gender = "Gender"
```

**Comment by Gouvernathor** [comment]
You can take the workarounds, but there is a bug here. The Toggle action does its job (and sets `renpy.display.behavior.current_input_value` and `renpy.display.behavior.input_value_active` to the good values), which even goes to enable the caret on the input, but _sometimes_ the input object doesn't intercept the keystroke event. For example, if you type "f" while the caret is blinking, sometimes it will add "f" to the displayed string, (sometimes it will do nothing if you have too many characters written because of the limit of 6,) but sometimes it will reach the keymap and toggle the fullscreen.
Debugging this will require diving in renpy.display.core.Interface.interact_core I think, and I'm not up for that right now.

**Comment by m-from-space** [comment]
> I'm not sure if this will help you @m-from-space , but this is how I allow for multiple inputs on one screen:

Thanks, but I am not sure how this will solve the issue of inputs not reacting. Is it because you only have one input inside the screen at a time? So you never encounter problems when typing inside the input?

> You can take the workarounds, but there is a bug here.

Nice that you taken time to investigate! Most often just activating an input and pressing "backspace" does not work at all, so it's not even about characters or something. And yes, it feels like the event is just not registered.

I have the same problem with `renpy.input()` as part of the normal script by the way. Sometimes it just doesn't react anymore. But it feels random to me.

**Comment by m-from-space** [comment]
> You can take the workarounds, but there is a bug here. 

I have an idea what is going on. The input field simply loses focus.

When you are inside an input field and you use the arrow keys (up and down), you are able to cycle through other buttons on the screen. If you do this often enough, the input loses focus, while the caret is still blinking. You can feel it because you then are not able to move the caret with left and right arrow keys anymore.

I'm not sure that arrow keys should even be considered when inside an input field.

The strange thing is, that sometimes I activate an input field and it immediately has no focus, although the caret is blinking.

**Comment by m-from-space** [comment]
This is not fixed for me in the latest nightly fix. But I encountered what the problem is: If you're enabling an input field, but then hover over another input field (or button for example), the active input field (still having a blinking caret) does not reply to your keystrokes anymore. If you hover over it again, it's reacting again. This is counter intuitive.

**Comment by renpytom** [comment]
It was fixed in the master branch. I just pushed the fix to nightly, so you'll get it there soon.

**Comment by m-from-space** [comment]
> It was fixed in the master branch. I just pushed the fix to nightly, so you'll get it there soon.

Thanks Tom! It works in general, which is awesome, but there are still two minor issues here:

If you have activated an input (as part of a button in my case), then hit "Return", the caret would still blink if your mouse is outside of the input. I would expect (as a user) that the input is disabled when "Return" is hit.

On the other hand, if your mouse is hovering another input while the caret is blinking and you hit "Return", the new input would be activated. I don't think that's a big deal though, just not something you would expect either.
